File: core/shared/src/main/scala/diplomacy/variant/standard/rule/OrderGenerator.scala
package diplomacy.variant.standard.rule

import diplomacy.rule.InvalidOrderMessage
import diplomacy.board.Power

trait OrderGenerator[Turn_ <: Turn, Power_ <: Power] extends Rule.TypeHelper {
  final type Turn = Turn_
  final type Power = Power_

  def ordersToSkipPhase(board: Board): Option[Set[Order]]
  def defaultOrder(board: Board)(unit: DiplomacyUnit): Order
}

File: src/standardRule/order-generator.ts
import { Unit, Board, ResolvedResult } from "./types"
import { Order } from "./order"

export interface OrderGenerator<Power> {
  ordersToSkipPhase (board: Board<Power>): Set<Order<Power>> | null
  /**
   * @return The default order of the unit. If there are no default order, it's null.
   */
  defaultOrderOf (board: Board<Power>, unit: Unit<Power>): Order<Power> | null
}
